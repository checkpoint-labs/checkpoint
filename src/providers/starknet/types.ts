import { RPC } from 'starknet';
import { BaseWriterParams } from '../../types';

// Shortcuts to starknet types.
export type Block = RPC.GetBlockWithTxs;
export type Transaction = RPC.GetBlockWithTxs['transactions'][number];
export type PendingTransaction = RPC.PendingTransactions[number];
export type Event = RPC.GetEventsResponse['events'][number];

// (Partially) narrowed types as real types are not exported from `starknet`.
export type FullBlock = Block & { block_number: number };
export type DeployTransaction = Transaction & { contract_address: string };

export type EventsMap = { [key: string]: Event[] };
export type ParsedEvent = Record<string, any>;

export type Writer = (
  args: {
    tx: Transaction;
    block: FullBlock | null;
    rawEvent?: Event;
    event?: ParsedEvent;
  } & BaseWriterParams
) => Promise<void>;

export function isFullBlock(block: Block): block is FullBlock {
  return 'block_number' in block;
}

export function isDeployTransaction(tx: Transaction | PendingTransaction): tx is DeployTransaction {
  return tx.type === 'DEPLOY';
}
