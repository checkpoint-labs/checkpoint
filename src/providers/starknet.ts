import { RpcProvider, hash, validateAndParseAddress } from 'starknet';
import { BaseProvider } from './base';
import { isFullBlock, isDeployTransaction } from '../types';
import type { Block, FullBlock, Transaction, Event, EventsMap } from '../types';

export class StarknetProvider extends BaseProvider {
  private readonly provider: RpcProvider;

  constructor({ instance, log }: ConstructorParameters<typeof BaseProvider>[0]) {
    super({ instance, log });

    this.provider = new RpcProvider({
      nodeUrl: this.instance.config.network_node_url
    });
  }

  async processBlock(blockNum: number) {
    let block: Block;
    let blockEvents: EventsMap;
    try {
      [block, blockEvents] = await Promise.all([
        this.provider.getBlockWithTxs(blockNum),
        this.getEvents(blockNum)
      ]);

      if (!isFullBlock(block) || block.block_number !== blockNum) {
        this.log.error({ blockNumber: blockNum }, 'invalid block');
        throw new Error('invalid block');
      }
    } catch (e) {
      if ((e as Error).message.includes('StarknetErrorCode.BLOCK_NOT_FOUND')) {
        this.log.info({ blockNumber: blockNum }, 'block not found');
      } else {
        this.log.error({ blockNumber: blockNum, err: e }, 'getting block failed... retrying');
      }

      throw e;
    }

    await this.handleBlock(block, blockEvents);

    await this.instance.setLastIndexedBlock(block.block_number);
  }

  private async handleBlock(block: FullBlock, blockEvents: EventsMap) {
    this.log.info({ blockNumber: block.block_number }, 'handling block');

    for (const [i, tx] of block.transactions.entries()) {
      await this.handleTx(
        block,
        i,
        tx,
        tx.transaction_hash ? blockEvents[tx.transaction_hash] || [] : []
      );
    }

    this.log.debug({ blockNumber: block.block_number }, 'handling block done');
  }

  private async handleTx(block: FullBlock, txIndex: number, tx: Transaction, events: Event[]) {
    this.log.debug({ txIndex }, 'handling transaction');

    const writerParams = this.instance.getWriterParams();

    if (this.instance.config.tx_fn) {
      await this.instance.writer[this.instance.config.tx_fn]({
        block,
        tx,
        ...writerParams
      });
    }

    for (const source of this.instance.config.sources || []) {
      let foundContractData = false;
      const contract = validateAndParseAddress(source.contract);

      if (
        isDeployTransaction(tx) &&
        source.deploy_fn &&
        contract === validateAndParseAddress(tx.contract_address)
      ) {
        foundContractData = true;
        this.log.info(
          { contract: source.contract, txType: tx.type, handlerFn: source.deploy_fn },
          'found deployment transaction'
        );

        await this.instance.writer[source.deploy_fn]({
          source,
          block,
          tx,
          ...writerParams
        });
      }

      for (const event of events) {
        if (contract === validateAndParseAddress(event.from_address)) {
          for (const sourceEvent of source.events) {
            if (`0x${hash.starknetKeccak(sourceEvent.name).toString('hex')}` === event.keys[0]) {
              foundContractData = true;
              this.log.info(
                { contract: source.contract, event: sourceEvent.name, handlerFn: sourceEvent.fn },
                'found contract event'
              );

              await this.instance.writer[sourceEvent.fn]({
                source,
                block,
                tx,
                event,
                ...writerParams
              });
            }
          }
        }
      }

      if (foundContractData) {
        await this.instance.insertCheckpoints([
          { blockNumber: block.block_number, contractAddress: source.contract }
        ]);
      }
    }

    this.log.debug({ txIndex }, 'handling transaction done');
  }

  private async getEvents(blockNumber: number): Promise<EventsMap> {
    const events: Event[] = [];

    let continuationToken: string | undefined;
    do {
      const result = await this.provider.getEvents({
        from_block: { block_number: blockNumber },
        to_block: { block_number: blockNumber },
        chunk_size: 1000,
        continuation_token: continuationToken
      });

      events.push(...result.events);

      continuationToken = result.continuation_token;
    } while (continuationToken);

    return events.reduce((acc, event) => {
      if (!acc[event.transaction_hash]) acc[event.transaction_hash] = [];

      acc[event.transaction_hash].push(event);

      return acc;
    }, {});
  }
}
