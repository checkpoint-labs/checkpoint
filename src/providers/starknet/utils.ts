import type { Abi } from 'starknet';

type StructAbi = {
  name: string;
  size: number;
  members: { name: string; type: string }[];
};

export const parseStruct = (
  type: string,
  data: string[],
  { current, structs }: { current: number; structs: Record<string, StructAbi> }
) => {
  const struct = structs[type];
  let structCurrent = current;

  return struct.members.reduce((output, field) => {
    if (structs[field.type]) {
      output[field.name] = parseStruct(field.type, data, { current: structCurrent, structs });
      structCurrent += structs[field.type].size;

      return output;
    }

    output[field.name] = data[structCurrent];
    structCurrent++;

    return output;
  }, {});
};

export const parseEvent = (abi: Abi, name: string, data: string[]): Record<string, any> => {
  const event = abi.find(el => el.name === name && el.type === 'event');
  if (!event) throw new Error('Unsupported event');

  const structs = abi
    .filter(el => el.type === 'struct')
    .reduce((acc, structAbi) => {
      acc[structAbi.name] = structAbi;
      return acc;
    }, {});

  let length = 0;
  let current = 0;

  return event.data.reduce((output, field) => {
    if (length > 0) {
      output[field.name] = data.slice(current, current + length);
      current += length;
      length = 0;

      return output;
    }

    if (structs[field.type]) {
      output[field.name] = parseStruct(field.type, data, { current, structs });
      current += structs[field.type].size;

      return output;
    }

    output[field.name] = data[current];
    if (field.name.endsWith('_len')) {
      length = parseInt(BigInt(data[current]).toString());
    }

    current++;

    return output;
  }, {});
};
