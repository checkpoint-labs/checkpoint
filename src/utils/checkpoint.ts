import objectHash from 'object-hash';
import { CheckpointConfig, ContractSourceConfig, ContractTemplate } from '../types';

export const getContractsFromConfig = (config: CheckpointConfig): string[] => {
  return (config.sources || []).map(source => source.contract);
};

const getHashableProperties = (config: ContractTemplate | ContractSourceConfig) => ({
  contract: 'contract' in config ? config.contract : undefined,
  start: 'start' in config ? config.start : undefined,
  events: (config.events || []).map(event => event.name)
});

export const getConfigChecksum = (config: CheckpointConfig): string => {
  const { tx_fn, global_events, sources, templates } = config;

  return objectHash(
    {
      tx_fn,
      global_events,
      sources: (sources || []).map(source => getHashableProperties(source)),
      templates: Object.fromEntries(
        Object.entries(templates || {}).map(([key, value]) => [key, getHashableProperties(value)])
      )
    },
    {
      unorderedArrays: true
    }
  );
};
