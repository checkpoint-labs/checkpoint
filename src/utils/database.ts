import pluralize from 'pluralize';

export const getTableName = (name: string) => {
  if (name === '_metadata') return '_metadatas';

  return pluralize(name);
};
