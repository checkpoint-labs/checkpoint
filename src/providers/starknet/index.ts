import { RpcProvider, hash, validateAndParseAddress } from 'starknet';
import { BaseProvider, BlockNotFoundError } from '../base';
import { parseEvent } from './utils';
import type { Abi } from 'starknet';
import { isFullBlock, isDeployTransaction } from '../../types';
import type {
  Block,
  FullBlock,
  Transaction,
  PendingTransaction,
  Event,
  EventsMap,
  ParsedEvent
} from '../../types';

export class StarknetProvider extends BaseProvider {
  private readonly provider: RpcProvider;
  private processedPoolTransactions = new Set();

  constructor({ instance, log, abis }: ConstructorParameters<typeof BaseProvider>[0]) {
    super({ instance, log, abis });

    this.provider = new RpcProvider({
      nodeUrl: this.instance.config.network_node_url
    });
  }

  async getNetworkIdentifier(): Promise<string> {
    const result = await this.provider.getChainId();
    return `starknet_${result}`;
  }

  async processBlock(blockNum: number) {
    let block: Block;
    let blockEvents: EventsMap;
    try {
      [block, blockEvents] = await Promise.all([
        this.provider.getBlockWithTxs(blockNum),
        this.getEvents(blockNum)
      ]);

      if (!isFullBlock(block) || block.block_number !== blockNum) {
        this.log.error({ blockNumber: blockNum }, 'invalid block');
        throw new Error('invalid block');
      }
    } catch (e) {
      if ((e as Error).message.includes('Block not found')) {
        this.log.info({ blockNumber: blockNum }, 'block not found');
        throw new BlockNotFoundError();
      }

      this.log.error({ blockNumber: blockNum, err: e }, 'getting block failed... retrying');
      throw e;
    }

    await this.handleBlock(block, blockEvents);

    await this.instance.setLastIndexedBlock(block.block_number);

    return blockNum + 1;
  }

  async processPool(blockNumber: number) {
    const txs = await this.provider.getPendingTransactions();
    const receipts = await Promise.all(
      txs.map(async tx => {
        if (!tx.transaction_hash) return null;

        try {
          return await this.provider.getTransactionReceipt(tx.transaction_hash);
        } catch (err) {
          this.log.warn(
            { transactionHash: tx.transaction_hash, err },
            'getting transaction receipt failed'
          );
          return null;
        }
      })
    );

    const txsWithReceipts = txs.filter((_, index) => receipts[index] !== null);
    const eventsMap = receipts.reduce((acc, receipt) => {
      if (receipt === null) return acc;

      acc[receipt.transaction_hash] = receipt.events;
      return acc;
    }, {});

    await this.handlePool(txsWithReceipts, eventsMap, blockNumber);
  }

  private async handleBlock(block: FullBlock, eventsMap: EventsMap) {
    this.log.info({ blockNumber: block.block_number }, 'handling block');

    const txsToCheck = block.transactions.filter(
      tx => !this.processedPoolTransactions.has(tx.transaction_hash)
    );

    for (const [i, tx] of txsToCheck.entries()) {
      await this.handleTx(
        block,
        block.block_number,
        i,
        tx,
        tx.transaction_hash ? eventsMap[tx.transaction_hash] || [] : []
      );
    }

    this.processedPoolTransactions.clear();

    this.log.debug({ blockNumber: block.block_number }, 'handling block done');
  }

  private async handlePool(txs: PendingTransaction[], eventsMap: EventsMap, blockNumber: number) {
    this.log.info('handling pool');

    const txsToCheck = txs.filter(tx => !this.processedPoolTransactions.has(tx.transaction_hash));

    for (const [i, tx] of txsToCheck.entries()) {
      await this.handleTx(
        null,
        blockNumber,
        i,
        tx,
        tx.transaction_hash ? eventsMap[tx.transaction_hash] || [] : []
      );

      this.processedPoolTransactions.add(tx.transaction_hash);
    }

    this.log.info('handling pool done');
  }

  private async handleTx(
    block: FullBlock | null,
    blockNumber: number,
    txIndex: number,
    tx: Transaction,
    events: Event[]
  ) {
    this.log.debug({ txIndex }, 'handling transaction');

    const writerParams = await this.instance.getWriterParams();

    if (this.instance.config.tx_fn) {
      await this.instance.writer[this.instance.config.tx_fn]({
        blockNumber,
        block,
        tx,
        ...writerParams
      });
    }

    if (this.instance.config.global_events) {
      const globalEventHandlers = this.instance.config.global_events.reduce((handlers, event) => {
        handlers[`0x${hash.starknetKeccak(event.name).toString('hex')}`] = {
          name: event.name,
          fn: event.fn
        };
        return handlers;
      }, {});

      for (const [eventIndex, event] of events.entries()) {
        const handler = globalEventHandlers[event.keys[0]];
        if (!handler) continue;

        this.log.info(
          { contract: event.from_address, event: handler.name, handlerFn: handler.fn },
          'found contract event'
        );

        await this.instance.writer[handler.fn]({
          block,
          blockNumber,
          tx,
          rawEvent: event,
          eventIndex,
          ...writerParams
        });
      }
    }

    for (const source of this.instance.config.sources || []) {
      let foundContractData = false;
      const contract = validateAndParseAddress(source.contract);

      if (
        isDeployTransaction(tx) &&
        source.deploy_fn &&
        contract === validateAndParseAddress(tx.contract_address)
      ) {
        foundContractData = true;
        this.log.info(
          { contract: source.contract, txType: tx.type, handlerFn: source.deploy_fn },
          'found deployment transaction'
        );

        await this.instance.writer[source.deploy_fn]({
          source,
          block,
          blockNumber,
          tx,
          ...writerParams
        });
      }

      for (const [eventIndex, event] of events.entries()) {
        if (contract === validateAndParseAddress(event.from_address)) {
          for (const sourceEvent of source.events) {
            if (`0x${hash.starknetKeccak(sourceEvent.name).toString('hex')}` === event.keys[0]) {
              foundContractData = true;
              this.log.info(
                { contract: source.contract, event: sourceEvent.name, handlerFn: sourceEvent.fn },
                'found contract event'
              );

              let parsedEvent: ParsedEvent | undefined;
              if (source.abi && this.abis?.[source.abi]) {
                try {
                  parsedEvent = parseEvent(
                    this.abis?.[source.abi] as Abi,
                    sourceEvent.name,
                    event.data
                  );
                } catch (err) {
                  this.log.warn(
                    { contract: source.contract, txType: tx.type, handlerFn: source.deploy_fn },
                    'failed to parse event'
                  );
                }
              }

              await this.instance.writer[sourceEvent.fn]({
                source,
                block,
                blockNumber,
                tx,
                rawEvent: event,
                event: parsedEvent,
                eventIndex,
                ...writerParams
              });
            }
          }
        }
      }

      if (foundContractData) {
        await this.instance.insertCheckpoints([{ blockNumber, contractAddress: source.contract }]);
      }
    }

    this.log.debug({ txIndex }, 'handling transaction done');
  }

  private async getEvents(blockNumber: number): Promise<EventsMap> {
    const events: Event[] = [];

    let continuationToken: string | undefined;
    do {
      const result = await this.provider.getEvents({
        from_block: { block_number: blockNumber },
        to_block: { block_number: blockNumber },
        chunk_size: 1000,
        continuation_token: continuationToken
      });

      events.push(...result.events);

      continuationToken = result.continuation_token;
    } while (continuationToken);

    return events.reduce((acc, event) => {
      if (!acc[event.transaction_hash]) acc[event.transaction_hash] = [];

      acc[event.transaction_hash].push(event);

      return acc;
    }, {});
  }
}
