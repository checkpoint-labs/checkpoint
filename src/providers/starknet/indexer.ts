import { Logger } from '../../utils/logger';
import { Instance, BaseIndexer } from '../base';
import { StarknetProvider } from '.';
import { Writer } from './types';

export class StarknetIndexer extends BaseIndexer {
  private writers: Record<string, Writer>;

  constructor(writers: Record<string, Writer>) {
    super();
    this.writers = writers;
  }

  init({ instance, log, abis }: { instance: Instance; log: Logger; abis?: Record<string, any> }) {
    this.provider = new StarknetProvider({ instance, log, abis, writers: this.writers });
  }

  public getHandlers(): string[] {
    return Object.keys(this.writers);
  }
}
