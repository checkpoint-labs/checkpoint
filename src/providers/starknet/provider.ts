import { RpcProvider, hash, validateAndParseAddress } from 'starknet';
import { BlockWithTxReceipts, SPEC } from '@starknet-io/types-js';
import { BaseProvider, BlockNotFoundError, ReorgDetectedError } from '../base';
import { parseEvent } from './utils';
import { CheckpointRecord } from '../../stores/checkpoints';
import { Block, FullBlock, Event, EventsData, ParsedEvent, isFullBlock, Writer } from './types';
import { ContractSourceConfig } from '../../types';
import { sleep } from '../../utils/helpers';

class CustomJsonRpcError extends Error {
  constructor(message: string, public code: number, public data: any) {
    super(message);
  }
}

export class StarknetProvider extends BaseProvider {
  private readonly provider: RpcProvider;
  private readonly writers: Record<string, Writer>;
  private seenPoolTransactions = new Set();
  private processedTransactions = new Set();
  private sourceHashes = new Map<string, string>();
  private logsCache = new Map<number, Event[]>();

  constructor({
    instance,
    log,
    abis,
    writers
  }: ConstructorParameters<typeof BaseProvider>[0] & {
    writers: Record<string, Writer>;
  }) {
    super({ instance, log, abis });

    this.provider = new RpcProvider({
      nodeUrl: this.instance.config.network_node_url
    });
    this.writers = writers;
  }

  formatAddresses(addresses: string[]): string[] {
    return addresses.map(address => validateAndParseAddress(address));
  }

  async getNetworkIdentifier(): Promise<string> {
    const result = await this.provider.getChainId();
    return `starknet_${result}`;
  }

  async getLatestBlockNumber(): Promise<number> {
    return this.provider.getBlockNumber();
  }

  async getBlockHash(blockNumber: number) {
    const block = await this.provider.getBlock(blockNumber);
    return block.block_hash;
  }

  async processBlock(blockNum: number, parentHash: string | null) {
    let block: Block | null = null;
    let eventsData: EventsData;
    try {
      const skipBlockFetching = this.instance.opts?.skipBlockFetching ?? false;
      const hasPreloadedBlockEvents = skipBlockFetching && this.logsCache.has(blockNum);

      if (!hasPreloadedBlockEvents) {
        block = await this.provider.getBlockWithTxHashes(blockNum);
      }

      if (block && (!isFullBlock(block) || block.block_number !== blockNum)) {
        this.log.error({ blockNumber: blockNum }, 'invalid block');
        throw new Error('invalid block');
      }

      eventsData = await this.getEvents({
        blockNumber: blockNum,
        blockHash: block?.block_hash ?? null
      });
    } catch (e) {
      if ((e as Error).message.includes('Block not found')) {
        this.log.info({ blockNumber: blockNum }, 'block not found');
        throw new BlockNotFoundError();
      }

      this.log.error({ blockNumber: blockNum, err: e }, 'getting block failed... retrying');
      throw e;
    }

    if (block && parentHash && block.parent_hash !== parentHash) {
      this.log.error({ blockNumber: blockNum }, 'reorg detected');
      throw new ReorgDetectedError();
    }

    await this.handleBlock(blockNum, block, eventsData);

    if (block && isFullBlock(block)) {
      await this.instance.setBlockHash(blockNum, block.block_hash);
    }

    await this.instance.setLastIndexedBlock(blockNum);

    return blockNum + 1;
  }

  async processPool(blockNumber: number) {
    const block = await this.getBlockWithReceipts('pending');
    const receipts = block.transactions
      .map(({ receipt }) => receipt)
      .filter(receipt => !this.seenPoolTransactions.has(receipt.transaction_hash));

    const txIds = receipts.map(receipt => receipt.transaction_hash);

    const eventsData = {
      isPreloaded: false,
      events: receipts.reduce((acc, receipt) => {
        if (receipt === null) return acc;

        acc[receipt.transaction_hash] = receipt.events;
        return acc;
      }, {})
    };

    await this.handlePool(txIds, eventsData, blockNumber);
  }

  private async handleBlock(blockNumber: number, block: FullBlock | null, eventsData: EventsData) {
    this.log.info({ blockNumber }, 'handling block');

    const blockTransactions = Object.keys(eventsData.events);
    const txsToCheck = blockTransactions.filter(txId => !this.seenPoolTransactions.has(txId));

    for (const txId of txsToCheck) {
      await this.handleTx(
        block,
        blockNumber,
        txId,
        eventsData.isPreloaded,
        eventsData.events[txId] || []
      );
    }

    this.seenPoolTransactions.clear();

    this.log.debug({ blockNumber }, 'handling block done');
  }

  private async handlePool(txIds: string[], eventsData: EventsData, blockNumber: number) {
    this.log.info('handling pool');

    for (const txId of txIds) {
      await this.handleTx(
        null,
        blockNumber,
        txId,
        eventsData.isPreloaded,
        eventsData.events[txId] || []
      );

      this.seenPoolTransactions.add(txId);
    }

    this.log.info('handling pool done');
  }

  private async handleTx(
    block: FullBlock | null,
    blockNumber: number,
    txId: string,
    isPreloaded: boolean,
    events: Event[]
  ) {
    this.log.debug({ txId }, 'handling transaction');

    if (this.processedTransactions.has(txId)) {
      this.log.warn({ hash: txId }, 'transaction already processed');
      return;
    }

    let wasTransactionProcessed = false;
    const helpers = await this.instance.getWriterHelpers();

    if (this.instance.config.tx_fn) {
      await this.writers[this.instance.config.tx_fn]({
        blockNumber,
        block,
        txId,
        helpers
      });

      wasTransactionProcessed = true;
    }

    if (this.instance.config.global_events) {
      const globalEventHandlers = this.instance.config.global_events.reduce((handlers, event) => {
        handlers[this.getEventHash(event.name)] = {
          name: event.name,
          fn: event.fn
        };
        return handlers;
      }, {});

      for (const event of events) {
        const handler = globalEventHandlers[event.keys[0]];
        if (!handler) continue;

        this.log.info(
          { contract: event.from_address, event: handler.name, handlerFn: handler.fn },
          'found contract event'
        );

        await this.writers[handler.fn]({
          block,
          blockNumber,
          txId,
          rawEvent: event,
          helpers
        });

        wasTransactionProcessed = true;
      }
    }

    let lastSources = this.instance.getCurrentSources(blockNumber);
    let sourcesQueue = [...lastSources];

    let source: ContractSourceConfig | undefined;
    while ((source = sourcesQueue.shift())) {
      let foundContractData = false;
      const contract = validateAndParseAddress(source.contract);

      for (const event of events) {
        if (contract === validateAndParseAddress(event.from_address)) {
          for (const sourceEvent of source.events) {
            if (this.getEventHash(sourceEvent.name) === event.keys[0]) {
              foundContractData = true;
              this.log.info(
                { contract: source.contract, event: sourceEvent.name, handlerFn: sourceEvent.fn },
                'found contract event'
              );

              let parsedEvent: ParsedEvent | undefined;
              if (source.abi && this.abis?.[source.abi]) {
                try {
                  parsedEvent = parseEvent(this.abis[source.abi], event);
                } catch (err) {
                  this.log.warn(
                    { contract: source.contract, txId, handlerFn: sourceEvent.fn },
                    'failed to parse event'
                  );
                }
              }

              await this.writers[sourceEvent.fn]({
                source,
                block,
                blockNumber,
                txId,
                rawEvent: event,
                event: parsedEvent,
                helpers
              });

              wasTransactionProcessed = true;
            }
          }
        }
      }

      if (wasTransactionProcessed) {
        this.processedTransactions.add(txId);
      }

      if (foundContractData) {
        await this.instance.insertCheckpoints([
          { blockNumber, contractAddress: validateAndParseAddress(source.contract) }
        ]);

        const nextSources = this.instance.getCurrentSources(blockNumber);
        const newSources = nextSources.slice(lastSources.length);

        sourcesQueue = sourcesQueue.concat(newSources);
        lastSources = this.instance.getCurrentSources(blockNumber);

        if (isPreloaded && newSources.length > 0) {
          this.handleNewSourceAdded();

          this.log.info(
            { newSources: newSources.map(s => s.contract) },
            'new sources added, fetching missing events'
          );

          const newSourcesEvents = await this.getEventsForSources({
            fromBlock: blockNumber,
            toBlock: blockNumber,
            sources: newSources
          });

          events = events.concat(newSourcesEvents);
        }
      }
    }

    this.log.debug({ txId }, 'handling transaction done');
  }

  private async getBlockWithReceipts(blockId: SPEC.BLOCK_ID) {
    const res = await fetch(this.instance.config.network_node_url, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json'
      },
      body: JSON.stringify({
        jsonrpc: '2.0',
        id: 1,
        method: 'starknet_getBlockWithReceipts',
        params: [blockId]
      })
    });

    if (!res.ok) {
      throw new Error(`Request failed: ${res.statusText}`);
    }

    const json = await res.json();

    if (json.error) {
      throw new CustomJsonRpcError(json.error.message, json.error.code, json.error.data);
    }

    return json.result as BlockWithTxReceipts;
  }

  private async getEvents({
    blockHash,
    blockNumber
  }: {
    blockHash: string | null;
    blockNumber: number;
  }): Promise<EventsData> {
    let isPreloaded = false;
    let events: Event[] = [];

    if (this.logsCache.has(blockNumber)) {
      isPreloaded = true;
      events = this.logsCache.get(blockNumber) || [];
      this.logsCache.delete(blockNumber);
    } else {
      if (!blockHash) {
        throw new Error('Block hash is required to fetch logs from network');
      }

      let continuationToken: string | undefined;
      do {
        const result = await this.provider.getEvents({
          from_block: { block_hash: blockHash },
          to_block: { block_hash: blockHash },
          chunk_size: 1000,
          continuation_token: continuationToken
        });

        events = events.concat(result.events);

        continuationToken = result.continuation_token;
      } while (continuationToken);
    }

    return {
      isPreloaded,
      events: events.reduce((acc, event) => {
        if (!acc[event.transaction_hash]) acc[event.transaction_hash] = [];

        acc[event.transaction_hash].push(event);

        return acc;
      }, {})
    };
  }

  async getEventsRangeForAddress(
    fromBlock: number,
    toBlock: number,
    address: string,
    eventNames: string[]
  ): Promise<Event[]> {
    let events: Event[] = [];

    let continuationToken: string | undefined;
    do {
      try {
        const result = await this.provider.getEvents({
          from_block: { block_number: fromBlock },
          to_block: { block_number: toBlock },
          address: address,
          keys: [eventNames.map(name => this.getEventHash(name))],
          chunk_size: 1000,
          continuation_token: continuationToken
        });

        events = events.concat(result.events);

        continuationToken = result.continuation_token;
      } catch (e) {
        this.log.error(
          { fromBlock, toBlock, continuationToken, address, err: e },
          'getEvents failed'
        );

        await sleep(5000);
      }
    } while (continuationToken);

    return events;
  }

  async getEventsForSources({
    fromBlock,
    toBlock,
    sources
  }: {
    fromBlock: number;
    toBlock: number;
    sources: ContractSourceConfig[];
  }): Promise<Event[]> {
    let events: Event[] = [];

    for (const source of sources) {
      const addressEvents = await this.getEventsRangeForAddress(
        fromBlock,
        toBlock,
        source.contract,
        source.events.map(event => event.name)
      );
      events = events.concat(addressEvents);
    }

    return events;
  }

  async getCheckpointsRange(fromBlock: number, toBlock: number): Promise<CheckpointRecord[]> {
    const events = await this.getEventsForSources({
      fromBlock,
      toBlock,
      sources: this.instance.getCurrentSources(fromBlock)
    });

    for (const log of events) {
      if (!this.logsCache.has(log.block_number)) {
        this.logsCache.set(log.block_number, []);
      }

      this.logsCache.get(log.block_number)?.push(log);
    }

    return events.map(event => ({
      blockNumber: event.block_number,
      contractAddress: validateAndParseAddress(event.from_address)
    }));
  }

  getEventHash(eventName: string) {
    if (!this.sourceHashes.has(eventName)) {
      this.sourceHashes.set(eventName, `0x${hash.starknetKeccak(eventName).toString(16)}`);
    }

    return this.sourceHashes.get(eventName) as string;
  }

  handleNewSourceAdded(): void {
    this.log.info('new source added, clearing logs cache');
    this.logsCache.clear();
  }
}
