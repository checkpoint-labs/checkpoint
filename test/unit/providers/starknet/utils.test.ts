import { parseEvent } from '../../../../src/providers/starknet/utils';
import {
  spaceFactoryAbi,
  spaceAbi,
  factoryAbiCairo1,
  spaceDeployedEvent,
  voteCreatedEvent,
  spaceDeployedEventCairo1
} from './fixtures';

describe('utils', () => {
  describe('parseEvent', () => {
    it('should parse event', () => {
      const output = parseEvent(spaceFactoryAbi, spaceDeployedEvent);

      expect(output).toMatchSnapshot();
    });

    it('should parse nested event', () => {
      const output = parseEvent(spaceAbi, voteCreatedEvent);

      expect(output).toMatchSnapshot();
    });

    it('should parse cairo 1 event', () => {
      const output = parseEvent(factoryAbiCairo1, spaceDeployedEventCairo1);

      expect(output).toMatchSnapshot();
    });
  });
});
