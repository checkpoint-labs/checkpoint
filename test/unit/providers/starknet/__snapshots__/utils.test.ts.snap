// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`utils parseEvent should parse event 1`] = `
{
  "authenticators": [
    "0x64cce9272197eba6353f5bbf060e097e516b411e66e83a9cf5910a08697df14",
    "0x4112e7aef90c47058238ccb76bf79ad5188afdf366870015185e3c7468ccbd9",
    "0x68a2d3c6d882ec0e2e94042556878d27e832a28e1308df04ad35fd8bae9ec6b",
    "0x64bb9fd620d7e4c5f5895329e8f1d3d5f485ccfb2b16345a1ca86658f24c9f6",
    "0x59283b509832027a386b3f419628a5b149e9d6462a6547c63e92bd4a09a7245",
    "0x5e1f273ca9a11f78bfb291cbe1b49294cf3c76dd48951e7ab7db6d9fb1e7d62",
  ],
  "authenticators_len": "0x6",
  "controller": "0x77d5345916a48b88ad002ead40344edbaa3a900d66c40d0b8e0c4c9f529d9a6",
  "deployer_address": "0x77d5345916a48b88ad002ead40344edbaa3a900d66c40d0b8e0c4c9f529d9a6",
  "execution_strategies": [
    "0x4ecc83848a519cc22b0d0ffb70e65ec8dde85d3d13439eff7145d4063cf6b4d",
    "0x21dda40770f4317582251cffd5a0202d6b223dc167e5c8db25dc887d11eba81",
  ],
  "execution_strategies_len": "0x2",
  "max_voting_duration": "0x15180",
  "metadata_uri": undefined,
  "metadata_uri_len": "0x0",
  "min_voting_duration": "0x0",
  "proposal_threshold": {
    "high": "0x0",
    "low": "0x1",
  },
  "quorum": {
    "high": "0x0",
    "low": "0x1",
  },
  "space_address": "0x3993631a1336a81c474e2cf5ebbb3e820547ad14e6f6a7afcdac9d35eaccac5",
  "voting_delay": "0x0",
  "voting_strategies": [
    "0x58623786b93d9b6ed1f83cec5c6fa6bea5f399d2795ee56a6123bdd83f5aa48",
    "0xd1b81feff3095ca9517fdfc7427e742ce96f7ca8f3b2664a21b2fba552493b",
  ],
  "voting_strategies_len": "0x2",
  "voting_strategy_params_flat": [
    "0x2",
    "0x0",
    "0x0",
    "0xb4fbf271143f5000000000000000000000000000",
    "0x3",
  ],
  "voting_strategy_params_flat_len": "0x5",
}
`;

exports[`utils parseEvent should parse nested event 1`] = `
{
  "proposal_id": "0xf",
  "vote": {
    "choice": "0x1",
    "voting_power": {
      "high": "0x0",
      "low": "0x23cdb7dc255bf4da",
    },
  },
  "voter_address": {
    "value": "0xef8305e140ac520225daf050e2f71d5fbcc543e7",
  },
}
`;
